primitives/blockchain/src/backend.rs

```diff
@@ -183,104 +183,51 @@ pub trait Backend<Block: BlockT>:
 	/// Return hashes of all blocks that are children of the block with `parent_hash`.
 	fn children(&self, parent_hash: Block::Hash) -> Result<Vec<Block::Hash>>;
 
-	/// Get the most recent block hash of the best (longest) chains
-	/// that contain block with the given `target_hash`.
+	/// Get the most recent block hash of the longest chain that contains
+	/// a block with the given `base_hash`.
 	///
 	/// The search space is always limited to blocks which are in the finalized
 	/// chain or descendents of it.
 	///
-	/// If `maybe_max_block_number` is `Some(max_block_number)`
-	/// the search is limited to block `numbers <= max_block_number`.
-	/// in other words as if there were no blocks greater `max_block_number`.
-	/// Returns `Ok(None)` if `target_hash` is not found in search space.
-	/// TODO: document time complexity of this, see [#1444](https://github.com/paritytech/substrate/issues/1444)
+	/// Returns `Ok(None)` if `base_hash` is not found in search space.
+	// TODO: document time complexity of this, see [#1444](https://github.com/paritytech/substrate/issues/1444)
 	fn best_containing(
 		&self,
-		target_hash: Block::Hash,
-		maybe_max_number: Option<NumberFor<Block>>,
+		base_hash: Block::Hash,
 		import_lock: &RwLock<()>,
 	) -> Result<Option<Block::Hash>> {
-		let target_header = {
-			match self.header(target_hash)? {
-				Some(x) => x,
-				// target not in blockchain
-				None => return Ok(None),
-			}
+		let Some(base_header) = self.header(base_hash)? else {
+			return Ok(None)
 		};
 
-		if let Some(max_number) = maybe_max_number {
-			// target outside search range
-			if target_header.number() > &max_number {
-				return Ok(None)
-			}
-		}
-
 		let leaves = {
 			// ensure no blocks are imported during this code block.
 			// an import could trigger a reorg which could change the canonical chain.
 			// we depend on the canonical chain staying the same during this code block.
 			let _import_guard = import_lock.read();
-
 			let info = self.info();
-
-			// this can be `None` if the best chain is shorter than the target header.
-			let maybe_canon_hash = self.hash(*target_header.number())?;
-
-			if maybe_canon_hash.as_ref() == Some(&target_hash) {
-				// if a `max_number` is given we try to fetch the block at the
-				// given depth, if it doesn't exist or `max_number` is not
-				// provided, we continue to search from all leaves below.
-				if let Some(max_number) = maybe_max_number {
-					if let Some(header) = self.hash(max_number)? {
-						return Ok(Some(header))
-					}
-				}
-			} else if info.finalized_number >= *target_header.number() {
-				// header is on a dead fork.
+			if info.finalized_number > *base_header.number() {
```

Comment:
Yeah we need to bring back the old check on if `hash(target_number) == target_hash`.

Reply:
This code restricts the search just blocks that are above the last finalized block.
This is guaranteed by the check at the begin, i.e. we proceed only if `finalized_number <= base.number`.
If not then we assume is on a "dead-fork" (aka a fork below finalization point).
(but note that - e.g. because of block pinning or archive mode - it can be on a dead fork even if is above finalization point).

The routine then, for each leaf and starting from the longest one, backtracks to check if at some point `base_hash` is hit. If from the current fork we can't reach the `base` then the current fork may be:
1. dead (probably not possible since IIRC `leaves()` doesn't return leaves of dead brances) OR 
2. not dead but starts from below the `base` that has been requested (thus not good anyway)

If in the end we can't find anything it may be because `base` is itself on a dead fork or for some *mysterious* reason (and in this case would be more appropriate to print the warning).
Unfortunately I suspect that, because of "archive mode", to be 100% sure that is on a dead fork (instead of the *mysterious* reason) we have to backtrack down to the first block we have... but I'm not sure if the isn't a more clever way.

Reply:
My comment was aiming at the comment in the code that says, search space includes the finalized chain. But it only includes blocks above the finalized chain. But again this is  pretty nitpicky.